worker/src/main/java/gov/cms/ab2d/worker/adapter/bluebutton/ContractBeneSearchImpl.java

```diff
@@ -0,0 +1,189 @@
+
+        long startTime = System.currentTimeMillis();
+        // Iterate through the futures
+        while (!futureHandles.isEmpty()) {
+            // See if there are any threads left to finish
+            Iterator<Future<ContractMapping>> iterator = futureHandles.iterator();
+            while (iterator.hasNext()) {
+                Future<ContractMapping> future = iterator.next();
+                // If it's done, claim the data and add it to the results, then remove it from the active threads
+                if (future.isDone()) {
+                    results.add(getContractMapping(future, contractNumber));
+                    iterator.remove();
+                }
+            }
+            // If we haven't removed all items from the running futures list, sleep for a bit
+            if (!futureHandles.isEmpty()) {
+                try {
+                    Thread.sleep(SLEEP_DURATION);
+                } catch (InterruptedException e) {
+                    log.warn("interrupted exception in thread.sleep(). Ignoring");
+                }
+            }
+        }
+        long endTime = System.currentTimeMillis();
+        int numMinutes = (int) (endTime - startTime) / 1000 / 60;
+        int totalRecords = results.stream().mapToInt(c -> c.getPatients().size()).sum();
+        eventLogger.log(new ReloadEvent(null, ReloadEvent.FileType.CONTRACT_MAPPING,
+                "Contract: " + contractNumber + " retrieved " + totalRecords + " in " + numMinutes, totalRecords));
+        // Return the finished results
+        return results;
+    }
+
+    /**
+     * Retrieve the data from the future after it's "done"
+     *
+     * @param future     - the future
+     * @param contractId - the contract number
+     * @return - the mapping
+     */
+    private ContractMapping getContractMapping(Future<ContractMapping> future, String contractId) throws ExecutionException, InterruptedException {
+        try {
+            return future.get();
+        } catch (InterruptedException | ExecutionException e) {
+            log.warn("InterruptedException while calling Future.get() - Getting Mapping for " + contractId);
+            throw e;
+        } catch (CancellationException e) {
+            // This could happen in the rare event that a job was cancelled mid-process.
+            // due to which the futures in the queue (that were not yet in progress) were cancelled.
+            // Nothing to be done here
+            log.warn("CancellationException while calling Future.get() - Getting Mapping for " + contractId);
+        }
+        return null;
+    }
+
+    private void addDateRangeToExistingOrNewPatient(ContractBeneficiaries beneficiaries,
+                                                    DateRange monthDateRange, String bfdPatientId) {
+        if (beneficiaries == null) {
+            return;
+        }
+        Optional<PatientDTO> optPatient = beneficiaries.getPatients().stream()
```

Comment:
This code may run into issues when there's a large amount of patients if it's searching through a list of possibly hundreds of thousands for each mapping, did you measure this on PDP-10000? It may be worth keeping a map of patient ids -> patients, or add the date range at an earlier point in time so that not all of these are being searched through.

Reply:
I can clean this up, but I don't think it will be a significant factor in your numbers so shouldn't hold up performance tests.

Reply:
I did some quick tests and it's probably still under a second to do all these for up to a million patients.

Reply:
Can you add a quick log message for timing between lines 60 and 68? When I did just the filtering iterating on a list of 200K in a test it took over 2 min.

Just something like 
long start = System.currentTimeMillis();
long end = System.currentTimeMillis();
log.info("Time to add date range to patients {}", end - start);

Reply:
Just saw this. Okay, done.